pilot/pkg/serviceregistry/serviceentry/servicediscovery.go

```diff
@@ -744,6 +734,13 @@ func servicesDiff(os []*model.Service, ns []*model.Service) ([]*model.Service, [
 	return added, deleted, updated, unchanged
 }
 
+func servicesEqual(os *model.Service, ns *model.Service) bool {
+	// TODO(ramaraochavali): do a field level comparison and exclude autoallocated addresses.
+	s := os.DeepCopy()
```

Comment:
no need deepcopy here, a shallow copy is ok


copied:=*os
copied.AutoAllocatedAddress = ""

Reply:
i tried - shallow copy is not working as it has locks and also some tests are failing with shallow copy

Reply:
I actually think DeepCopy is not costly here as we do only during add/updates - not during gets.

Reply:
With the patch below, i can not find any failure in servicediscovery_test.go
```
--- a/pilot/pkg/serviceregistry/serviceentry/servicediscovery.go
+++ b/pilot/pkg/serviceregistry/serviceentry/servicediscovery.go
@@ -728,7 +728,7 @@ func servicesDiff(os []*model.Service, ns []*model.Service) ([]*model.Service, [
                newSvc, f := newServiceHosts[s.Hostname]
                if !f {
                        deleted = append(deleted, s)
-               } else if !reflect.DeepEqual(s, newSvc) {
+               } else if !servicesEqual(s, newSvc) {
                        updated = append(updated, newSvc)
                } else {
                        unchanged = append(unchanged, newSvc)
@@ -744,6 +744,12 @@ func servicesDiff(os []*model.Service, ns []*model.Service) ([]*model.Service, [
        return added, deleted, updated, unchanged
 }

+func servicesEqual(os *model.Service, ns *model.Service) bool {
+       s := *os
+       s.AutoAllocatedAddress = ""
+       return reflect.DeepEqual(&s, ns)
+}
+

```

Reply:
Strange... It failed some tests and now it passes.